blog/2021-07-30-temporal-tips-tricks-1.md

```diff
@@ -0,0 +1,116 @@
+```
+
+Obviously constantly rerunning a command on every file save may be a problem in some cases, so use your best judgement.
+
+## Code Snippets
+
+This is a generally applicable developer productivity tip, but there will always be some amount of unavoidable boilerplate in code and Temporal is no exception. Roll your commonly used code into reusable snippets, and take advantage of modern snippet features like [Tab Stops, Placeholders, and Transformations](https://macromates.com/manual/en/snippets#tab_stops).
+
+```go
+// @prefix newclient
+// @description add new client
+
+// The client is a heavyweight object that should be created once
+// options https://pkg.go.dev/go.temporal.io/sdk@v1.8.0/internal#ClientOptions
+${0:newClient}, err := client.NewClient(client.Options{})
+if err != nil {
+	log.Fatalln("Unable to create client", err)
+}
+defer ${0:newClient}.Close()
+```
+
+You can even leave comments to relevant documentation, like this:
+
+```go
+// @prefix RetryPolicy
+// @description simple retry policy
+
+// https://docs.temporal.io/docs/go/retries/
+    RetryPolicy: 			&temporal.RetryPolicy{
+			InitialInterval:    time.Second,
+			BackoffCoefficient: 2,
+			MaximumInterval:    time.Minute * 10,
+			MaximumAttempts:    5,
+		},
+```
+
+You can look up how to do this in [TextMate](https://macromates.com/manual/en/snippets), [Atom](https://flight-manual.atom.io/using-atom/sections/snippets/), and [Sublime Text](http://docs.sublimetext.info/en/latest/extensibility/snippets.html). 
+
+In VS Code, you should try [the Easy Snippet extension](https://marketplace.visualstudio.com/items?itemName=inu1255.easy-snippet) to make writing [VS Code's User Defined Snippets](https://code.visualstudio.com/docs/editor/userdefinedsnippets) easier. Here's a [short video](https://twitter.com/swyx/status/1420772267772968960) demonstrating usage of Easy Snippet.
+
+## Stack Trace
+
+One of the lesser known but extremely useful features of [Temporal Web](https://docs.temporal.io/docs/system-tools/web-ui#viewing-stack-traces-on-temporal-web) is the ability to view stack traces.
+
+![https://s3-us-west-2.amazonaws.com/secure.notion-static.com/be6f3198-a73c-46f6-84ef-8623b12be710/Untitled.png](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/be6f3198-a73c-46f6-84ef-8623b12be710/Untitled.png)
+
+This helps greatly with debugging and understanding how Temporal deals with asynchronous work and is able to resume workflows on a different machine.
+
+## Event History
+
+You can export the event history via `[tctl](https://docs.temporal.io/docs/system-tools/tctl#show-workflow-history)` or the [Web UI](https://docs.temporal.io/docs/system-tools/web-ui#execution-histories-on-temporal-web):
+
+```bash
+tctl workflow show \
+  -w myWorkflowId \
+  -r 866ae14c-88cf-4f1e-980f-571e031d71b0
+```
+
+This is not just good for recordkeeping, but also great for migrating Workflows to new versions *while* they are still running!
```

Comment:
What does this have to do with new versions?

Reply:
```suggestion
This is not just good for recordkeeping, but also great for taking a failing production workflow and debug it locally, exactly as it was executed.
```